test/functional/test-viewer.js

```diff
@@ -450,4 +450,42 @@ describe('Viewer', () => {
       });
     });
   });
+
+  describe('viewerOrigin', () => {
+
+    it('should return empty string if origin is not known', () => {
+      const viewer = new Viewer(windowApi);
+      return viewer.getViewerOrigin().then(viewerOrigin => {
+        expect(viewerOrigin).to.equal('');
+      });
+    });
+
+    it('should return ancestor origin if known', () => {
+      windowApi.parent = {};
+      windowApi.location.ancestorOrigins = ['https://google.com'];
+      const viewer = new Viewer(windowApi);
+      return viewer.getViewerOrigin().then(viewerOrigin => {
+        expect(viewerOrigin).to.equal('https://google.com');
+      });
+    });
+
+    it('should return viewer origin if set via handshake', () => {
+      windowApi.parent = {};
+      const viewer = new Viewer(windowApi);
+      const result = viewer.getViewerOrigin().then(viewerOrigin => {
+        expect(viewerOrigin).to.equal('https://foobar.com');
+      });
+      viewer.setMessageDeliverer(() => {}, 'https://foobar.com');
+      return result;
+    });
+
+    it('should return empty string if handshake does not happen', () => {
+      windowApi.parent = {};
```

Comment:
Please use fakeClock here - we don't want the test to actually wait for 1 second.